File: Trees/257 Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        res = []
        def preorder(root,l):
            if root: 
                if (root.left==None and root.right==None):
                    res.append(l+[root.val])
                else:
                    preorder(root.left,l+[root.val])
                    preorder(root.right,l+[root.val])
                
        preorder(root,[])
        return [('->'.join([str(k) for k in i]))for i in res]
            

File: Trees/257 Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> res;
    void preorder(TreeNode* root,vector<int> ans)
    {
        if(root)
        {
            if (root->left==NULL && root->right==NULL)
            {
                ans.push_back(root->val);
                res.push_back(ans);
            }
            else
            {
                ans.push_back(root->val);
                preorder(root->left,ans);
                preorder(root->right,ans);
            }
        }
    }
    vector<string> binaryTreePaths(TreeNode* root) 
    {
        vector<int> ans;
        preorder(root,ans);
        vector<string> ans1;

        for (int i = 0; i < res.size(); i++)
        {
            string s="";
            int j;
            for (j = 0; j < res[i].size()-1; j++)
            {
                s+=to_string(res[i][j])+"->";
            }
            s+=to_string(res[i][j++]);
            ans1.push_back(s);
        }
        return ans1;
    }
};
